### PR TITLE
Decouple the uploading of files from the finding of changed files.

### PR DESCRIPTION
**Pull request by hcarver**
This solves a problem where file modifications aren't uploaded the first time around, and are then lost.

Instead we maintain a queue of paths to be uploaded, with paths removed from the queue only when upload is successful.

### CODE REVIEW

**Comment by hcarver** [comment]
FYI, I tested that two changes to the same file do not double the upload frequency when uploads are failing (they don't), and that uploads suceeding stop that upload from being retried (they do).

File: content-watcher.js [hunk 1]
```diff
@@ -34,31 +36,52 @@ function sendPing() {
...
+        process.stdout.write(d)
+        process.stdout.write("\n")
+      })
+
+      // 1xx and 2xx status codes are successful for our purposes
+      if(res.statusCode < 300) {
+        this_index = changed_files.indexOf(path)
+        if(this_index >= 0) {
+          changed_files.splice(this_index, 1)
+        }
```
Comment (davidmillican) [inline content-watcher.js]: I often wish Javascript could have better built-in methods...

File: content-watcher.js [hunk 2]
```diff
@@ -34,31 +36,52 @@ function sendPing() {
...
+    const options = postRequestOptions(
+      `/attendances/${process.env.ATTENDANCE_ID}/file_snapshots`,
+      {
+        "Content-Type": "application/json",
+        "Content-Length": data.length
+      }
+    )
+
+    const req = https.request(options, res => {
+      process.stdout.write(`status: ${res.statusCode}\n`)
```
Comment (davidmillican) [inline content-watcher.js]: Why was this change needed?

File: content-watcher.js [hunk 3]
```diff
@@ -93,10 +116,18 @@ const pollDirectoryForChanges = dirPath => {
 
 
 const pollerFunction = () => {
+  // Set the timeout first, to ensure an exception doesn't stop the recursion
+  setTimeout(pollerFunction, 1000)
```
Comment (davidmillican) [inline content-watcher.js:124]: This should probably have been a separate commit - it seems useful on its own.

File: content-watcher.js [hunk 4]
```diff
@@ -34,31 +36,52 @@ function sendPing() {
...
-  const req = https.request(options, res => {
-    console.log(`status: ${res.statusCode}`)
+function uploadChangedFiles() {
+  // Iterate over a copy so the indexes don't change as we modify the array
+  changed_files_copy = [...changed_files]
 
-    res.on("data", d => {
-      process.stdout.write(d)
+  changed_files_copy.forEach(function(path) {
+    process.stdout.write(`uploading: ${path}\n`)
```
Comment (davidmillican) [inline content-watcher.js]: This `forEach` block ends up being quite long - just for readability, I'd probably want to pull it out into a separate function. Up to you though.

**Review by davidmillican — APPROVED** [review]
Looks good.

File: content-watcher.js (hunk 1 above)
Comment (hcarver) [reply content-watcher.js]: Goodness me yes.

File: content-watcher.js (hunk 2 above)
Comment (hcarver) [reply content-watcher.js]: Consistency with the other output lines. I'm not sure it's needed though.

File: content-watcher.js (hunk 3 above)
Comment (hcarver) [reply content-watcher.js:124]: Yeah, that's true.
